File: packages/mcp-server/src/tools/WriteLogs.ts
import fs from "fs/promises";
import path from "path";
import { DateTime } from "luxon";
import { z } from "zod";
import type { McpServer } from "../server.js";
import type { ToolContext } from "../types.js";
import {
  formatISOWeekDate,
  getDayAbbreviation,
  parseISOWeekDate,
  findDaySection,
} from "../utils/log-format.js";

/**
 * Replace an entire day section in the log with new entries
 * Pass an empty object to delete the entire day section
 */
export async function writeLogs(
  logPath: string,
  isoWeekDate: string,
  entries: Record<string, string>
): Promise<{ count: number; errors: string[]; deleted: boolean }> {
  const errors: string[] = [];

  // Validate and parse ISO week date
  const dayDateTime = parseISOWeekDate(isoWeekDate);
  if (!dayDateTime) {
    errors.push(
      `Invalid ISO week date format: '${isoWeekDate}'. Expected format: YYYY-Www-D (e.g., 2025-W50-1)`
    );
    return { count: 0, errors, deleted: false };
  }

  const dayAbbrev = getDayAbbreviation(dayDateTime);
  const dayHeader = `## ${isoWeekDate} (${dayAbbrev})`;

  // Handle empty object - delete the entire day section
  if (Object.keys(entries).length === 0) {
    // Read existing log content
    let logContent = "";
    try {
      logContent = await fs.readFile(logPath, "utf-8");
    } catch (error) {
      // File doesn't exist - nothing to delete
      return { count: 0, errors: [], deleted: false };
    }

    // Find and remove the day section
    const lines = logContent.split("\n");
    const section = findDaySection(lines, dayHeader);

    if (!section) {
      // Day section doesn't exist - nothing to delete
      return { count: 0, errors: [], deleted: false };
    }

    // Remove the section
    lines.splice(section.start, section.end - section.start);

    // Clean up extra blank lines
    const updatedContent = lines.join("\n").replace(/\n{3,}/g, "\n\n");
    await fs.writeFile(logPath, updatedContent);

    return { count: 0, errors: [], deleted: true };
  }

  // Parse and sort entries by time
  const parsedEntries: Array<{ time: DateTime; timeStr: string; message: string }> =
    [];

  for (const [timeStr, message] of Object.entries(entries)) {
    const parsed = DateTime.fromFormat(timeStr, "h:mm a");
    if (!parsed.isValid) {
      errors.push(
        `Invalid time format: '${timeStr}'. Expected format: h:mm AM/PM (e.g., 9:00 AM, 2:30 PM)`
      );
      continue;
    }

    parsedEntries.push({ time: parsed, timeStr, message });
  }

  // If there were parsing errors, return early
  if (errors.length > 0) {
    return { count: 0, errors, deleted: false };
  }

  // Sort chronologically
  parsedEntries.sort((a, b) => a.time.toMillis() - b.time.toMillis());

  // Generate the new day section
  const entryLines = parsedEntries.map(
    (entry) => `- ${entry.timeStr} – ${entry.message}`
  );
  const newSection = [dayHeader, "", ...entryLines, ""].join("\n");

  // Read existing log content
  let logContent = "";
  try {
    logContent = await fs.readFile(logPath, "utf-8");
  } catch (error) {
    // File doesn't exist, create it with just this section
    await fs.writeFile(logPath, newSection);
    return { count: parsedEntries.length, errors: [], deleted: false };
  }

  // Find and replace the day section
  const lines = logContent.split("\n");
  const section = findDaySection(lines, dayHeader);

  if (!section) {
    // Day section doesn't exist - append at the end
    const updatedContent = logContent.endsWith("\n")
      ? logContent + "\n" + newSection
      : logContent + "\n\n" + newSection;
    await fs.writeFile(logPath, updatedContent);
  } else {
    // Replace the existing section
    lines.splice(section.start, section.end - section.start, ...newSection.split("\n"));
    await fs.writeFile(logPath, lines.join("\n"));
  }

  return { count: parsedEntries.length, errors: [], deleted: false };
}

/**
 * WriteLogs Tool
 *
 * Replace an entire day's log entries in Log.md with consolidated/compacted entries.
 * Use this ONLY during memory consolidation when you need to rewrite or summarize a day's logs.
 * For adding new entries during active work, use the Log tool instead.
 */
export function registerWriteLogs(server: McpServer, context: ToolContext) {
  server.registerTool(
    "WriteLogs",
    {
      title: "Rewrite Day's Log Entries",
      description:
        "Replace an entire day's log entries with consolidated/compacted entries. " +
        "Use this ONLY during memory consolidation to rewrite or summarize a day's logs. " +
        "For adding new entries during active work, use the Log tool instead (it's simpler and doesn't require reading the log first). " +
        "This tool automatically formats entries with correct timestamps, en-dashes, and chronological sorting. " +
        "Pass an empty object to delete the entire day section (header and all entries).",
      inputSchema: {
        isoWeekDate: z
          .string()
          .describe(
            "ISO week date in YYYY-Www-D format (e.g., '2025-W50-1' for Monday of week 50). " +
              "Week starts on Monday (1=Mon, 7=Sun)."
          ),
        entries: z
          .record(z.string(), z.string())
          .describe(
            "Object mapping time strings to log messages. " +
              "Keys: '9:00 AM', '2:30 PM', etc. (12-hour format with AM/PM). " +
              "Values: Log entry content. " +
              "Example: { '9:00 AM': 'Started investigation', '2:30 PM': 'Fixed bug #123' }"
          ),
      },
      annotations: {
        readOnlyHint: false,
        destructiveHint: true, // Replaces entire day section
        openWorldHint: false,
      },
    },
    async ({ isoWeekDate, entries }) => {
      const { vaultPath } = context;
      const logPath = path.join(vaultPath, "Log.md");

      const { count, errors, deleted } = await writeLogs(logPath, isoWeekDate, entries);

      if (errors.length > 0) {
        return {
          content: [
            {
              type: "text",
              text: `Failed to write logs:\n${errors.map((e) => `- ${e}`).join("\n")}`,
            },
          ],
          isError: true,
        };
      }

      if (deleted) {
        return {
          content: [
            {
              type: "text",
              text: `Deleted day section ${isoWeekDate}`,
            },
          ],
          structuredContent: {
            isoWeekDate,
            deleted: true,
          },
        };
      }

      return {
        content: [
          {
            type: "text",
            text: `Successfully wrote ${count} log entries for ${isoWeekDate}`,
          },
        ],
        structuredContent: {
          isoWeekDate,
          count,
        },
      };
    }
  );
}

File: crates/mcp-server/src/tools/write_logs.rs
//! WriteLogs tool for bulk replacing a day's log entries.
//!
//! Used during memory consolidation to rewrite or summarize a day's logs.

use rmcp::model::{CallToolResult, Content, ErrorData};
use std::collections::HashMap;
use std::path::Path;
use tokio::fs;

use super::log::get_day_abbreviation_from_iso;

/// Replace an entire day's log entries with new entries.
pub async fn execute(
    vault_path: &Path,
    iso_week_date: &str,
    entries: HashMap<String, String>,
) -> Result<CallToolResult, ErrorData> {
    // Validate ISO week date format
    if !is_valid_iso_week_date(iso_week_date) {
        return Err(ErrorData::invalid_params(
            format!(
                "Invalid ISO week date format: '{}'. Expected format: YYYY-Www-D (e.g., '2025-W50-1')",
                iso_week_date
            ),
            None,
        ));
    }

    // Validate all time entries
    let mut invalid_times = Vec::new();
    for time in entries.keys() {
        if parse_time_12h(time).is_none() {
            invalid_times.push(time.clone());
        }
    }

    if !invalid_times.is_empty() {
        return Err(ErrorData::invalid_params(
            format!(
                "Invalid time format(s): {}. Expected format: 'h:mm AM' or 'h:mm PM'",
                invalid_times.join(", ")
            ),
            None,
        ));
    }

    let log_path = vault_path.join("Log.md");
    let day_abbrev = get_day_abbreviation_from_iso(iso_week_date);
    let day_header = format!("## {} ({})", iso_week_date, day_abbrev);

    // Read existing content
    let log_content = match fs::read_to_string(&log_path).await {
        Ok(content) => content,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => String::new(),
        Err(e) => {
            return Err(ErrorData::internal_error(
                format!("Failed to read Log.md: {}", e),
                None,
            ));
        }
    };

    let mut lines: Vec<String> = log_content.lines().map(String::from).collect();

    // Find the section for this day
    let section_start = lines.iter().position(|l| l == &day_header);

    // If entries is empty, delete the entire day section
    if entries.is_empty() {
        if let Some(start) = section_start {
            // Find the end of this section (next ## header or end of file)
            let mut end = start + 1;
            while end < lines.len() {
                if lines[end].starts_with("##") {
                    break;
                }
                end += 1;
            }

            // Remove the section
            lines.drain(start..end);

            // Clean up extra blank lines
            cleanup_blank_lines(&mut lines);

            let content = lines.join("\n");
            if let Err(e) = fs::write(&log_path, content).await {
                return Err(ErrorData::internal_error(
                    format!("Failed to write Log.md: {}", e),
                    None,
                ));
            }

            return Ok(CallToolResult::success(vec![Content::text(format!(
                "Deleted day section for {}",
                iso_week_date
            ))]));
        } else {
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "No entries to delete - day section {} does not exist",
                iso_week_date
            ))]));
        }
    }

    // Sort entries chronologically
    let mut sorted_entries: Vec<(String, String)> = entries.into_iter().collect();
    sorted_entries.sort_by(|a, b| {
        let time_a = parse_time_12h(&a.0).unwrap_or((24, 0));
        let time_b = parse_time_12h(&b.0).unwrap_or((24, 0));
        time_a.cmp(&time_b)
    });

    // Format new entries
    let new_entries: Vec<String> = sorted_entries
        .iter()
        .map(|(time, message)| format!("- {} – {}", time, message))
        .collect();

    if let Some(start) = section_start {
        // Find the end of this section
        let mut end = start + 1;
        while end < lines.len() {
            if lines[end].starts_with("##") {
                break;
            }
            end += 1;
        }

        // Replace the section content (keeping header)
        let mut new_section = vec![day_header, String::new()];
        new_section.extend(new_entries);
        new_section.push(String::new());

        lines.splice(start..end, new_section);
    } else {
        // Add new section at the end
        if !lines.is_empty() && !lines.last().map(|l| l.is_empty()).unwrap_or(true) {
            lines.push(String::new());
        }
        lines.push(day_header);
        lines.push(String::new());
        lines.extend(new_entries);
        lines.push(String::new());
    }

    let content = lines.join("\n");
    if let Err(e) = fs::write(&log_path, content).await {
        return Err(ErrorData::internal_error(
            format!("Failed to write Log.md: {}", e),
            None,
        ));
    }

    Ok(CallToolResult::success(vec![Content::text(format!(
        "Replaced {} entries for {}",
        sorted_entries.len(),
        iso_week_date
    ))]))
}

/// Validate ISO week date format: YYYY-Www-D
fn is_valid_iso_week_date(s: &str) -> bool {
    // Check format: YYYY-Www-D
    let parts: Vec<&str> = s.split('-').collect();
    if parts.len() != 3 {
        return false;
    }

    // Year: 4 digits
    if parts[0].len() != 4 || parts[0].parse::<u32>().is_err() {
        return false;
    }

    // Week: Www where w is 01-53
    if !parts[1].starts_with('W') || parts[1].len() != 3 {
        return false;
    }
    match parts[1][1..].parse::<u32>() {
        Ok(w) if (1..=53).contains(&w) => {}
        _ => return false,
    };

    // Day: 1-7
    match parts[2].parse::<u32>() {
        Ok(d) if (1..=7).contains(&d) => {}
        _ => return false,
    };

    true
}

/// Parse 12-hour time format (e.g., "9:30 AM") to (hour24, minute)
fn parse_time_12h(s: &str) -> Option<(u32, u32)> {
    let parts: Vec<&str> = s.split_whitespace().collect();
    if parts.len() != 2 {
        return None;
    }

    let time_parts: Vec<&str> = parts[0].split(':').collect();
    if time_parts.len() != 2 {
        return None;
    }

    let hour: u32 = time_parts[0].parse().ok()?;
    let minute: u32 = time_parts[1].parse().ok()?;

    if hour < 1 || hour > 12 || minute > 59 {
        return None;
    }

    let am_pm = parts[1].to_uppercase();
    let hour_24 = match am_pm.as_str() {
        "AM" if hour == 12 => 0,
        "AM" => hour,
        "PM" if hour == 12 => 12,
        "PM" => hour + 12,
        _ => return None,
    };

    Some((hour_24, minute))
}

/// Clean up multiple consecutive blank lines to at most 2
fn cleanup_blank_lines(lines: &mut Vec<String>) {
    let mut i = 0;
    let mut blank_count = 0;

    while i < lines.len() {
        if lines[i].trim().is_empty() {
            blank_count += 1;
            if blank_count > 2 {
                lines.remove(i);
                continue;
            }
        } else {
            blank_count = 0;
        }
        i += 1;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn test_is_valid_iso_week_date() {
        assert!(is_valid_iso_week_date("2025-W50-1"));
        assert!(is_valid_iso_week_date("2025-W01-7"));
        assert!(is_valid_iso_week_date("2026-W52-3"));

        assert!(!is_valid_iso_week_date("2025-50-1")); // Missing W
        assert!(!is_valid_iso_week_date("2025-W54-1")); // Invalid week
        assert!(!is_valid_iso_week_date("2025-W50-8")); // Invalid day
        assert!(!is_valid_iso_week_date("2025-W50-0")); // Invalid day
        assert!(!is_valid_iso_week_date("invalid"));
    }

    #[test]
    fn test_parse_time_12h() {
        assert_eq!(parse_time_12h("9:30 AM"), Some((9, 30)));
        assert_eq!(parse_time_12h("12:00 PM"), Some((12, 0)));
        assert_eq!(parse_time_12h("12:00 AM"), Some((0, 0)));
        assert_eq!(parse_time_12h("3:45 PM"), Some((15, 45)));
        assert_eq!(parse_time_12h("11:59 PM"), Some((23, 59)));

        assert_eq!(parse_time_12h("13:00 PM"), None); // Invalid hour
        assert_eq!(parse_time_12h("9:60 AM"), None); // Invalid minute
        assert_eq!(parse_time_12h("9:30"), None); // Missing AM/PM
        assert_eq!(parse_time_12h("invalid"), None);
    }

    #[tokio::test]
    async fn test_write_logs_creates_new_section() {
        let temp_dir = TempDir::new().unwrap();
        let vault_path = temp_dir.path();
        let log_path = vault_path.join("Log.md");

        // Create empty log file
        fs::write(&log_path, "").await.unwrap();

        let mut entries = HashMap::new();
        entries.insert("9:00 AM".to_string(), "Started work".to_string());
        entries.insert("2:30 PM".to_string(), "Finished task".to_string());

        let result = execute(vault_path, "2025-W50-1", entries).await;
        assert!(result.is_ok());

        let content = fs::read_to_string(&log_path).await.unwrap();
        assert!(content.contains("## 2025-W50-1 (Mon)"));
        assert!(content.contains("9:00 AM – Started work"));
        assert!(content.contains("2:30 PM – Finished task"));
    }

    #[tokio::test]
    async fn test_write_logs_replaces_existing_section() {
        let temp_dir = TempDir::new().unwrap();
        let vault_path = temp_dir.path();
        let log_path = vault_path.join("Log.md");

        // Create log with existing content
        let initial_content = "## 2025-W50-1 (Mon)\n\n- 8:00 AM – Old entry\n\n";
        fs::write(&log_path, initial_content).await.unwrap();

        let mut entries = HashMap::new();
        entries.insert("10:00 AM".to_string(), "New entry".to_string());

        let result = execute(vault_path, "2025-W50-1", entries).await;
        assert!(result.is_ok());

        let content = fs::read_to_string(&log_path).await.unwrap();
        assert!(!content.contains("Old entry"));
        assert!(content.contains("New entry"));
    }

    #[tokio::test]
    async fn test_write_logs_deletes_section_when_empty() {
        let temp_dir = TempDir::new().unwrap();
        let vault_path = temp_dir.path();
        let log_path = vault_path.join("Log.md");

        // Create log with content
        let initial_content = "## 2025-W50-1 (Mon)\n\n- 8:00 AM – Entry\n\n";
        fs::write(&log_path, initial_content).await.unwrap();

        let entries = HashMap::new(); // Empty = delete

        let result = execute(vault_path, "2025-W50-1", entries).await;
        assert!(result.is_ok());

        let content = fs::read_to_string(&log_path).await.unwrap();
        assert!(!content.contains("2025-W50-1"));
    }

    #[tokio::test]
    async fn test_write_logs_invalid_iso_date() {
        let temp_dir = TempDir::new().unwrap();
        let vault_path = temp_dir.path();

        let entries = HashMap::new();
        let result = execute(vault_path, "invalid", entries).await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_write_logs_invalid_time_format() {
        let temp_dir = TempDir::new().unwrap();
        let vault_path = temp_dir.path();

        let mut entries = HashMap::new();
        entries.insert("invalid".to_string(), "Message".to_string());

        let result = execute(vault_path, "2025-W50-1", entries).await;
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_write_logs_sorts_chronologically() {
        let temp_dir = TempDir::new().unwrap();
        let vault_path = temp_dir.path();
        let log_path = vault_path.join("Log.md");

        fs::write(&log_path, "").await.unwrap();

        let mut entries = HashMap::new();
        entries.insert("3:00 PM".to_string(), "Third".to_string());
        entries.insert("9:00 AM".to_string(), "First".to_string());
        entries.insert("12:00 PM".to_string(), "Second".to_string());

        execute(vault_path, "2025-W50-1", entries).await.unwrap();

        let content = fs::read_to_string(&log_path).await.unwrap();
        let first_pos = content.find("First").unwrap();
        let second_pos = content.find("Second").unwrap();
        let third_pos = content.find("Third").unwrap();

        assert!(first_pos < second_pos);
        assert!(second_pos < third_pos);
    }
}
